crowbar_framework/lib/crowbar/deployment_queue.rb

```diff
@@ -0,0 +1,410 @@
+        # is an uninteresting optimization.
+        delay, pre_cached_nodes = add_pending_elements(bc, inst, elements, queue_me)
+
+        # We have all nodes ready.
+        if delay.empty?
+          # There's a path: process_queue -> proposal_commit -> apply_role ->
+          # queue_proposal, which seems to be used as a test if all dependencies
+          # (queue_me = false) and nodes (delay.empty?) are still in that state
+          # by the time we want to apply. So if that is the case, we just drop
+          # proposal from the queue and exit.
+          #
+          # remove from queue if it was queued before; might not be in the queue
+          # because the proposal got changed since it got added to the queue
+          unless preexisting_queued_item.nil?
+            logger.debug("queue proposal: dequeuing already queued #{inst} #{bc}")
+            dequeued = dequeue_proposal_no_lock(db["proposal_queue"], inst, bc)
+            db.save if dequeued
+          end
+
+          return [ delay, pre_cached_nodes ]
+        end
+
+        # Delay not empty, we're missing some nodes.
+        # And proposal is not in queue
+        if preexisting_queued_item.nil?
+          db["proposal_queue"] << { "barclamp" => bc, "inst" => inst, "elements" => elements, "deps" => deps }
+        else
+          # Update (overwrite) item that is already in queue
+          preexisting_queued_item["elements"] = elements
+          preexisting_queued_item["deps"] = deps
+        end
+
+        db.save
+      rescue StandardError => e
+        logger.error("Error queuing proposal for #{bc}:#{inst}: #{e.message}")
+      ensure
+        file_lock.release(f)
+      end
+
+      # Mark the proposal as in the queue
+      prop = Proposal.where(barclamp: bc, name: inst).first
+      prop["deployment"][bc]["crowbar-queued"] = true
+      prop.save
+      logger.debug("queue proposal: exit #{inst} #{bc}")
+      [ delay, pre_cached_nodes ]
+    end
+
+    # Locking wrapper around dequeue_proposal_no_lock
+    def dequeue_proposal(bc, inst)
+      logger.debug("dequeue proposal: enter #{inst} #{bc}")
+      ret = false
+      begin
+        f = file_lock.acquire("queue", logger: logger)
+
+        db = Chef::DataBag.load("crowbar/queue") rescue nil
+        logger.debug("dequeue proposal: exit #{inst} #{bc}: no entry") if db.nil?
+        return [200, {}] if db.nil?
+
+        queue = db["proposal_queue"]
+        dequeued = dequeue_proposal_no_lock(queue, inst, bc)
+        db.save if dequeued
+      rescue StandardError => e
+        logger.error("Error dequeuing proposal for #{bc}:#{inst}: #{e.message} #{e.backtrace.join("\n")}")
+        logger.debug("dequeue proposal: exit #{inst} #{bc}: error")
+        return [400, e.message]
+      ensure
+        file_lock.release(f)
+      end
+      logger.debug("dequeue proposal: exit #{inst} #{bc}")
+      return dequeued ? [200, {}] : [400, '']
```

Comment:
Style/RedundantReturn: Redundant `return` detected.<br>Style/StringLiterals: Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.